### PR TITLE
Implement Token Tracking System

### PR DESCRIPTION
**Pull request by fabioberger**
This PR:
- Makes it such that each user only tracks WETH and ZRX by default and can add/remove tokens they wish to track.
- Only fetch balance/allowances for tracked tokens, so loading the dApp is much faster.
- Ask the user to confirm tracking an untracked token if selected in order generate form, or if trying to fill an order involving untracked tokens (could be up to two).
- Nest tracked tokens in local storage under userAddress and networkId.
- Separate balance/allowances out of Token type and into TokenState type as a step towards lazy-loading balances/allowances in a future PR.
- Add track and untrack buttons to the balances page.
- Fixed bug where unlocking metamask wouldn't get the token balances/allowances to be refetched.

screen shot 2017-09-13 at 15 13 05

screen shot 2017-09-13 at 15 13 17

screen shot 2017-09-13 at 19 52 08

screen shot 2017-09-13 at 19 52 13

### CODE REVIEW

File: ts/components/generate_order/new_token_form.tsx [hunk 1]
```diff
@@ -164,12 +164,16 @@ export class NewTokenForm extends React.Component<NewTokenFormProps, NewTokenFor
     private onTokenNameChanged(e: any, name: string) {
         let nameErrText = '';
         const maxLength = 30;
+        const tokens = _.values(this.props.tokenByAddress);
+        const tokenWithNameExists = !_.isUndefined(_.find(tokens, {name}));
```
Comment (LogvinovLeon) [inline ts/components/generate_order/new_token_form.tsx]: every field in object literal on it's own line ;)

File: ts/components/generate_order/asset_picker.tsx [hunk 2]
```diff
@@ -261,6 +261,21 @@ export class AssetPicker extends React.Component<AssetPickerProps, AssetPickerSt
         const tokenAddress = this.state.chosenTrackTokenAddress;
         const token = this.props.tokenByAddress[tokenAddress];
         const newTokenEntry = _.assign({}, token);
+
+        // Let's make sure this token (with a unique address), also have a unique name/symbol, otherwise
+        // we append something to make it visibly clear to the end user that this is a different underlying
+        // token then the identically named one they already had locally.
+        const existingTokens = _.values(this.props.tokenByAddress);
+        const isUniqueName = _.isUndefined(_.find(existingTokens, {name: newTokenEntry.name}));
```
Comment (LogvinovLeon) [inline ts/components/generate_order/asset_picker.tsx]: Assign find result to a var

File: ts/components/generate_order/asset_picker.tsx [hunk 3]
```diff
@@ -261,6 +261,21 @@ export class AssetPicker extends React.Component<AssetPickerProps, AssetPickerSt
...
+        if (!isUniqueName) {
+            newTokenEntry.name = `${newTokenEntry.name} [Imported]`;
```
Comment (LogvinovLeon) [inline ts/components/generate_order/asset_picker.tsx]: Assign [Imported] to a var and name it

File: ts/components/generate_order/asset_picker.tsx [hunk 4]
```diff
@@ -261,6 +261,21 @@ export class AssetPicker extends React.Component<AssetPickerProps, AssetPickerSt
...
+        }
+
+        const isUniqueSymbol = _.isUndefined(_.find(existingTokens, {symbol: newTokenEntry.symbol}));
```
Comment (LogvinovLeon) [inline ts/components/generate_order/asset_picker.tsx]: same

File: ts/components/dialogs/track_token_confirmation_dialog.tsx [hunk 5]
```diff
@@ -71,9 +71,25 @@ export class TrackTokenConfirmationDialog extends
...
             const newTokenEntry = _.assign({}, token);
+
+            if (!token.isRegistered) {
+                // Let's make sure this token (with a unique address), also have a unique name/symbol, otherwise
+                // we append something to make it visibly clear to the end user that this is a different underlying
+                // token then the identically named one they already had locally.
+                const existingTokens = _.values(this.props.tokenByAddress);
+                const isUniqueName = _.isUndefined(_.find(existingTokens, {name: newTokenEntry.name}));
+                if (!isUniqueName) {
+                    newTokenEntry.name = `${newTokenEntry.name} [Imported]`;
```
Comment (LogvinovLeon) [inline ts/components/dialogs/track_token_confirmation_dialog.tsx]: Can we add a wiki page with this behaviour documented? We append different custom prefixes to token names.
I don't like this, Maybe we should store this in state.

File: ts/components/dialogs/track_token_confirmation_dialog.tsx [hunk 6]
```diff
@@ -95,4 +111,7 @@ export class TrackTokenConfirmationDialog extends
         });
         this.props.onToggleDialog(didUserAcceptTracking);
     }
+    private addSymbolFlourish(symbol: string) {
+        return `*${symbol}*`;
```
Comment (LogvinovLeon) [inline ts/components/dialogs/track_token_confirmation_dialog.tsx]: What is this? Why is it needed?

File: ts/components/generate_order/asset_picker.tsx (hunk 3 above)
Comment (fabioberger) [reply ts/components/generate_order/asset_picker.tsx]: Outdated.

File: ts/components/generate_order/asset_picker.tsx [hunk 7]
```diff
@@ -111,6 +119,7 @@ export class AssetPicker extends React.Component<AssetPickerProps, AssetPickerSt
             <TrackTokenConfirmation
                 tokens={[token]}
                 networkId={this.props.networkId}
+                isAddingTokenToTracked={this.state.isAddingTokenToTracked}
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply ts/components/generate_order/asset_picker.tsx]: I prefer a more verbose name.

File: ts/components/generate_order/asset_picker.tsx [hunk 8]
```diff
@@ -1,50 +1,151 @@
...
+                            Tracking a token adds it to the balances section of 0x Portal and
+                            allows you to generate orders that involve this tokens. Would you
+                            like to start tracking this token?
+                        </div>
+                    </div>
+                }
+            </div>
+        );
+    }
+    private renderAssetPicker() {
```
[the remark this answers is not in the document]
Comment (fabioberger) [reply ts/components/generate_order/asset_picker.tsx]: Eventually. I different PR.

File: ts/components/dialogs/track_token_confirmation_dialog.tsx (hunk 5 above)
Comment (fabioberger) [reply ts/components/dialogs/track_token_confirmation_dialog.tsx]: outdated.

**Pull request merged**